### PR TITLE
MM-68501 - implement GetMaskedVisualAST and wire API handler

### PR DESCRIPTION
**Pull request by pvev**
#### Summary
Implements the read-path for the attribute value masking feature so when a admin opens a policy in the editor they only see the attribute values they themselves hold or public ones, not the full list configured in the policy.

The PR implements GetMaskedVisualAST in the app layer, which converts a CEL expression to a Visual AST and then filters each condition's values to only what the requesting caller holds. `shared_only` fields show the caller's own values, source_only fields are fully redacted, and public fields pass through untouched. The POST `/access_control_policies/cel/visual_ast` handler is wired to call this path when the AttributeValueMasking flag is on, with fail-closed behavior on unknown fields. HasMaskedValues is set on any condition where values were omitted so the client can render the appropriate UI affordance.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-68501

#### Screenshots
n/a

#### Release Note
```release-note
NONE
```

## Change Impact: 🟡 Medium

**Regression Risk:** Changes touch access-control and property-access logic across multiple modules (API handler, app masking, property validation, and value read paths), introducing new masking semantics and altering validation behavior; this creates moderate risk if uncommon configurations or edge CEL expressions aren't covered by tests. The implementation uses fail-closed semantics to avoid leaking attribute values, and substantial unit tests were added, but interactions at integration boundaries (store lookups, feature-flag gating, and API wiring) remain areas where regressions could surface.

**QA Recommendation:** Prioritize manual QA for: (1) feature-flag gating—verify POST /access_control_policies/cel/visual_ast uses masked vs. unmasked paths correctly; (2) masking behavior—confirm shared_only, source_only, and public semantics across select/multiselect/text/date/user fields (including binary masking for text fields); (3) validation/compatibility—check property field validation and rejection of contradictory configs (e.g., shared_only with member-writable permission values); and (4) fail-closed and error handling—exercise missing fields, corrupt values, and malformed CEL expressions. Given the added extensive unit tests, focus manual testing on integration flows and edge cases.

*Generated by CodeRabbitAI*

[Review Change Stack](https://app.coderabbit.ai/change-stack/mattermost/mattermost/pull/36413)

### CODE REVIEW

File: server/channels/app/access_control_masking_test.go [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+		assert.True(t, condition.HasMaskedValues)
+	})
+}
+
+func TestMaskConditionValues(t *testing.T) {
+	rctx := request.TestContext(t)
+
+	// nil App is safe for every branch that does not reach a.getCallerTextValues
+	// (i.e., everything except shared_only + text field, which needs a real store).
+	var a *App
```
[the remark this answers is not in the document]
Comment (pvev) [reply server/channels/app/access_control_masking_test.go:381]: done

File: server/channels/app/access_control_masking_test.go [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+
+	// If filterConditionValues were called on an AttrValue condition, it would
+	// incorrectly mask the attribute path since it won't be in the visible set.
+	filterConditionValues(condition, map[string]struct{}{})
+	assert.Nil(t, condition.Value)
+	assert.True(t, condition.HasMaskedValues)
+
+	// This confirms maskConditionValues MUST return early for AttrValue before
+	// reaching filterConditionValues — covered by integration tests via GetMaskedVisualAST.
+}
```
[the remark this answers is not in the document]
Comment (pvev) [reply server/channels/app/access_control_masking_test.go:263]: done

File: server/i18n/en.json [hunk 3]
```diff
@@ -7532,6 +7532,10 @@
     "id": "app.pap.get_channel_members_to_remove.app_error",
     "translation": "Could not get channel members to remove."
   },
+  {
+    "id": "app.pap.get_masked_visual_ast.app_error",
+    "translation": "Could not generate masked visual AST from expression."
+  },
```
[the remark this answers is not in the document]
Comment (pvev) [reply server/i18n/en.json:7582]: done

File: server/channels/app/properties/access_control.go [hunk 4]
```diff
@@ -1060,6 +1060,7 @@ func (pas *PropertyAccessService) filterSharedOnlyValue(field *model.PropertyFie
 // applyFieldReadAccessControl applies read access control to a single field.
 // Returns the field with options filtered based on the caller's access permissions.
 // - Public fields: returned as-is
+// - User-editable fields (PermissionValues=member): returned as-is so users can see all choices
```
Comment (davidkrauser) [inline server/channels/app/properties/access_control.go:1078]: **question:** is this a use case we want to support? Should we reject this configuration on field creation or update? I don't love having `PermissionValue`-related logic at this level (`PermissionValue` in an API-level gate). WDYT?

File: server/public/model/property_access.go [hunk 5]
```diff
@@ -59,13 +59,6 @@ func ValidatePropertyFieldAccessMode(field *PropertyField) error {
 		return fmt.Errorf("invalid access mode '%s'", accessMode)
 	}
 
-	// Validate shared_only is only used with select/multiselect fields
-	if accessMode == PropertyAccessModeSharedOnly {
```
Comment (davidkrauser) [inline server/public/model/property_access.go:63]: **question:** why remove this block? Do we have a use case for `shared_only` with other types?

File: server/channels/app/access_control.go [hunk 6]
```diff
@@ -480,6 +481,204 @@ func (a *App) publishChannelPolicyEnforcedUpdate(rctx request.CTX, channelID str
 	a.Publish(messageWs)
 }
 
+func (a *App) GetMaskedVisualAST(rctx request.CTX, expression string, callerID string) (*model.VisualExpression, *model.AppError) {
```
Comment (davidkrauser) [inline server/channels/app/access_control.go]: **issue:** this file (`access_control.go`) feels like a strange place to put this logic. Should we put it in a file with a more specific name and purpose (e.g. something related to policy masking?).

File: server/channels/app/access_control.go [hunk 7]
```diff
@@ -480,6 +481,204 @@ func (a *App) publishChannelPolicyEnforcedUpdate(rctx request.CTX, channelID str
...
+		return model.PropertyAccessModePublic
+	}
+	accessMode, ok := field.Attrs[model.PropertyAttrsAccessMode].(string)
+	if !ok {
+		return model.PropertyAccessModePublic
+	}
+	return accessMode
+}
+
+func extractVisibleOptionNames(field *model.PropertyField) map[string]struct{} {
```
Comment (davidkrauser) [inline server/channels/app/access_control.go]: **suggestion:** `extractFieldName`, `getFieldAccessMode`, and `extractVisibleOptionNames` all feel like helpers that would be useful elsewhere. I think we probably _do_ have very similar code elsewhere. Should we extract these out into generic helpers?

File: server/channels/app/access_control.go [hunk 8]
```diff
@@ -480,6 +481,204 @@ func (a *App) publishChannelPolicyEnforcedUpdate(rctx request.CTX, channelID str
...
+		name, ok := optMap["name"].(string)
+		if ok && name != "" {
+			names[name] = struct{}{}
+		}
+	}
+
+	return names
+}
+
+func (a *App) getCallerTextValues(rctx request.CTX, callerID string, field *model.PropertyField, cpaGroupID string) map[string]struct{} {
```
Comment (davidkrauser) [inline server/channels/app/access_control.go]: **question:** if we don't support access control on non-select fields, we won't need this method, right?

**Review by davidkrauser — COMMENTED** [review]
Looks good overall, just had some questions about opening this up to non-select/multiselect field types, and a few other minor things.

File: server/channels/app/properties/access_control.go (hunk 4 above)
Comment (pvev) [reply server/channels/app/properties/access_control.go:1078]: Agree. So, that escape hatch was a workaround for a configuration that is contradictory. Fixed by rejecting the combination at validation time in `ValidatePropertyFieldAccessMode` and removing the escape hatch. The contradiction is now a configuration error caught at field creation/update, and the service layer no longer carries permission level logic.

File: server/public/model/property_access.go (hunk 5 above)
Comment (pvev) [reply server/public/model/property_access.go:63]: ok, so, yes,`shared_only` on text fields are a real use case in LDAP/SAML synced attributes (program codenames, etc) these could are commonly stored as free form text values whose existence is controlled/senstive information. And this is now relevant since now system admins are also expected to hold the values shown in the policy or get masked values.

File: server/channels/app/access_control.go (hunk 6 above)
Comment (pvev) [reply server/channels/app/access_control.go]: agree and fixed in last commit.

File: server/channels/app/access_control.go (hunk 7 above)
Comment (pvev) [reply server/channels/app/access_control.go]: ok, thanks for the suggestion, so, I made some changes like:

- getFieldAccessMode  moved it onto *PropertyField as GetAccessMode() in property_access.go. The access mode is a property of the field itself, not really a masking action/function                                                                                                   
- extractFieldName kept in the masking file. It just parses the user.attributes. prefix off a CEL reference. If someone else needs the same parsing later we can promote it to a CEL utils package
- extractVisibleOptionNames  also kept in the masking file, is tightly coupled there.

File: server/channels/app/access_control.go (hunk 8 above)
Comment (pvev) [reply server/channels/app/access_control.go]: following response to your previous comment, yes, we do now, we need it.

File: server/channels/app/properties/access_control_field_test.go [hunk 9]
```diff
@@ -1441,35 +1441,6 @@ func TestLinkedPropertyField_SecurityInheritance(t *testing.T) {
...
-	// The member-writable guard must return the field as-is so the user can self-assign.
-	result := pas.applyFieldReadAccessControl(field, "some-user-id")
-	require.NotNil(t, result)
-	resultOptions, ok := result.Attrs[model.PropertyFieldAttributeOptions].([]any)
-	require.True(t, ok)
-	assert.Len(t, resultOptions, 3, "member-writable shared_only field must bypass option filtering")
-}
+// The previous "member-writable shared_only" early-return in applyFieldReadAccessControl
+// was removed in favor of rejecting that contradictory configuration at validation time
+// (see TestValidatePropertyFieldAccessMode in server/public/model/property_access_test.go).
```
Comment (davidkrauser) [inline server/channels/app/properties/access_control_field_test.go:1446]: Can we drop this comment?

File: server/public/model/property_access.go [hunk 10]
```diff
@@ -66,5 +79,13 @@ func ValidatePropertyFieldAccessMode(field *PropertyField) error {
 		}
 	}
 
+	// shared_only + member-writable is contradictory: shared_only filters what
+	// callers see to values they hold, but member-writable lets users self-assign
+	// any value. Reject the combination at validation time instead of working
+	// around it at the API/service layer.
+	if accessMode == PropertyAccessModeSharedOnly && field.PermissionValues != nil && *field.PermissionValues == PermissionLevelMember {
+		return fmt.Errorf("access mode 'shared_only' is incompatible with member-writable permission_values")
+	}
```
Comment (davidkrauser) [inline server/public/model/property_access.go:88]: Thinking about this more - I think this is right. If I could set my own value, I could brute force may way to seeing everyone else's value. Thanks for making the change 👍

**Review by davidkrauser — APPROVED** [review]
Nice work - just have one nitpick.

File: server/channels/app/properties/access_control_field_test.go (hunk 9 above)
Comment (pvev) [reply server/channels/app/properties/access_control_field_test.go:1446]: sure, will do in one of the next upcoming PRS.

Comment (pvev) [reply server/channels/app/properties/access_control_field_test.go:1446]: done: https://github.com/mattermost/mattermost/pull/36517/changes/7d9631e5655c90e41604042e17e06b3a0fa8a3e4

**Review by hmohammed-prodsec — APPROVED** [review]
